File: mobile/lib/features/navigation/services/routing_service.dart
import 'package:campusar/core/constants/app_constants.dart';
import 'package:campusar/core/utilities/geo_utils.dart';
import 'package:campusar/data/models/route_graph.dart';

class RouteNotFoundException implements Exception {
  RouteNotFoundException(this.message);
  final String message;
  @override
  String toString() => message;
}

class RouteCalculationResult {
  const RouteCalculationResult({
    required this.nodes,
    required this.totalDistanceMeters,
    required this.estimatedWalkingTimeSeconds,
  });

  final List<RouteNode> nodes;
  final double totalDistanceMeters;
  final double estimatedWalkingTimeSeconds;

  double get estimatedWalkingTimeMinutes => estimatedWalkingTimeSeconds / 60;
}

class _Adjacency {
  const _Adjacency(this.neighborId, this.edge);
  final String neighborId;
  final RouteEdge edge;
}

/// Runs A* over the campus pedestrian graph, entirely on-device, so
/// routing works with no network connection (section 13 / 17 of the
/// product spec). This mirrors backend/app/services/routing_service.py's
/// algorithm and edge-weighting exactly, so an offline route and a
/// server-calculated one for the same start/end agree.
class RoutingService {
  RoutingService(List<RouteNode> nodes, List<RouteEdge> edges) {
    for (final n in nodes) {
      _nodes[n.id] = n;
      _adjacency.putIfAbsent(n.id, () => []);
    }
    for (final e in edges) {
      // Pedestrian edges are bidirectional.
      _adjacency.putIfAbsent(e.fromNode, () => []).add(_Adjacency(e.toNode, e));
      _adjacency.putIfAbsent(e.toNode, () => []).add(_Adjacency(e.fromNode, e));
    }
  }

  final Map<String, RouteNode> _nodes = {};
  final Map<String, List<_Adjacency>> _adjacency = {};

  String nearestNode(double latitude, double longitude) {
    if (_nodes.isEmpty) {
      throw RouteNotFoundException('Routing graph is empty.');
    }
    String? bestId;
    double bestDistance = double.infinity;
    for (final node in _nodes.values) {
      final d = GeoUtils.haversineDistanceMeters(
        latitude,
        longitude,
        node.latitude,
        node.longitude,
      );
      if (d < bestDistance) {
        bestDistance = d;
        bestId = node.id;
      }
    }
    return bestId!;
  }

  RouteCalculationResult findRoute({
    required double fromLatitude,
    required double fromLongitude,
    required String toNodeId,
    bool accessibleOnly = false,
    double walkingSpeedMetersPerSecond =
        AppConstants.averageWalkingSpeedMetersPerSecond,
  }) {
    if (!_nodes.containsKey(toNodeId)) {
      throw RouteNotFoundException("Destination node '$toNodeId' not found.");
    }
    final startId = nearestNode(fromLatitude, fromLongitude);
    final nodeIds = _aStar(startId, toNodeId, accessibleOnly);
    final totalDistance = _pathDistance(nodeIds);

    return RouteCalculationResult(
      nodes: nodeIds.map((id) => _nodes[id]!).toList(),
      totalDistanceMeters: double.parse(totalDistance.toStringAsFixed(1)),
      estimatedWalkingTimeSeconds: double.parse(
        GeoUtils.estimateWalkingTimeSeconds(
          totalDistance,
          walkingSpeedMetersPerSecond: walkingSpeedMetersPerSecond,
        ).toStringAsFixed(1),
      ),
    );
  }

  double _heuristic(String nodeId, String goalId) {
    final a = _nodes[nodeId]!;
    final b = _nodes[goalId]!;
    return GeoUtils.haversineDistanceMeters(
      a.latitude,
      a.longitude,
      b.latitude,
      b.longitude,
    );
  }

  List<String> _aStar(String startId, String goalId, bool accessibleOnly) {
    final openHeap = HeapPriorityQueue<_ScoredNode>();
    openHeap.add(_ScoredNode(0, startId));

    final cameFrom = <String, String>{};
    final gScore = <String, double>{startId: 0};
    final visited = <String>{};

    while (openHeap.isNotEmpty) {
      final current = openHeap.removeFirst().nodeId;
      if (current == goalId) {
        return _reconstructPath(cameFrom, current);
      }
      if (visited.contains(current)) continue;
      visited.add(current);

      for (final adj in _adjacency[current] ?? const <_Adjacency>[]) {
        if (accessibleOnly && !adj.edge.accessible) continue;
        final tentativeG = gScore[current]! + adj.edge.distance;
        if (tentativeG < (gScore[adj.neighborId] ?? double.infinity)) {
          cameFrom[adj.neighborId] = current;
          gScore[adj.neighborId] = tentativeG;
          final fScore = tentativeG + _heuristic(adj.neighborId, goalId);
          openHeap.add(_ScoredNode(fScore, adj.neighborId));
        }
      }
    }

    throw RouteNotFoundException(
      "No route found between '$startId' and '$goalId'"
      '${accessibleOnly ? " using only accessible paths." : "."}',
    );
  }

  List<String> _reconstructPath(Map<String, String> cameFrom, String current) {
    final path = <String>[current];
    var node = current;
    while (cameFrom.containsKey(node)) {
      node = cameFrom[node]!;
      path.add(node);
    }
    return path.reversed.toList();
  }

  double _pathDistance(List<String> nodeIds) {
    double total = 0;
    for (var i = 0; i < nodeIds.length - 1; i++) {
      final a = nodeIds[i];
      final b = nodeIds[i + 1];
      final edge = (_adjacency[a] ?? const <_Adjacency>[])
          .firstWhere((adj) => adj.neighborId == b);
      total += edge.edge.distance;
    }
    return total;
  }
}

class _ScoredNode implements Comparable<_ScoredNode> {
  const _ScoredNode(this.score, this.nodeId);
  final double score;
  final String nodeId;

  @override
  int compareTo(_ScoredNode other) => score.compareTo(other.score);
}

/// Minimal binary-heap priority queue (avoids pulling in an extra package
/// just for A*'s open set).
class HeapPriorityQueue<T extends Comparable<T>> {
  final List<T> _heap = [];

  bool get isNotEmpty => _heap.isNotEmpty;

  void add(T value) {
    _heap.add(value);
    _bubbleUp(_heap.length - 1);
  }

  T removeFirst() {
    final result = _heap.first;
    final last = _heap.removeLast();
    if (_heap.isNotEmpty) {
      _heap[0] = last;
      _bubbleDown(0);
    }
    return result;
  }

  void _bubbleUp(int index) {
    while (index > 0) {
      final parent = (index - 1) ~/ 2;
      if (_heap[index].compareTo(_heap[parent]) >= 0) break;
      _swap(index, parent);
      index = parent;
    }
  }

  void _bubbleDown(int index) {
    final length = _heap.length;
    while (true) {
      final left = 2 * index + 1;
      final right = 2 * index + 2;
      var smallest = index;
      if (left < length && _heap[left].compareTo(_heap[smallest]) < 0) {
        smallest = left;
      }
      if (right < length && _heap[right].compareTo(_heap[smallest]) < 0) {
        smallest = right;
      }
      if (smallest == index) break;
      _swap(index, smallest);
      index = smallest;
    }
  }

  void _swap(int i, int j) {
    final tmp = _heap[i];
    _heap[i] = _heap[j];
    _heap[j] = tmp;
  }
}

File: backend/app/services/routing_service.py
"""
Campus pedestrian routing engine.

Implements A* shortest-path search over the RouteNode/RouteEdge graph
(see section 13 of the product spec). This is intentionally a simple,
swappable implementation: RoutingService only depends on plain node/edge
data structures, not on the database or web framework, so it can later be
replaced by a more sophisticated routing engine (e.g. one that accounts
for elevation, crowding, or indoor floor transitions) without touching
the API layer.
"""
from __future__ import annotations

import heapq
from dataclasses import dataclass, field

from app.core.geo import estimate_walking_time_seconds, haversine_distance_meters


class RouteNotFoundError(Exception):
    """Raised when no path exists between two nodes (e.g. disconnected graph,
    or accessible_only=True with no accessible path available)."""


@dataclass
class GraphNode:
    id: str
    latitude: float
    longitude: float
    type: str


@dataclass
class GraphEdge:
    id: str
    from_node: str
    to_node: str
    distance: float
    accessible: bool = True
    indoor: bool = False
    outdoor: bool = True


@dataclass
class RouteResult:
    node_ids: list[str]
    total_distance_meters: float
    estimated_walking_time_seconds: float


@dataclass
class RoutingGraph:
    """In-memory adjacency representation built from RouteNode/RouteEdge rows."""

    nodes: dict[str, GraphNode] = field(default_factory=dict)
    # adjacency[node_id] = list of (neighbor_id, edge)
    adjacency: dict[str, list[tuple[str, GraphEdge]]] = field(default_factory=dict)

    @classmethod
    def build(cls, nodes: list[GraphNode], edges: list[GraphEdge]) -> "RoutingGraph":
        graph = cls()
        for n in nodes:
            graph.nodes[n.id] = n
            graph.adjacency.setdefault(n.id, [])
        for e in edges:
            # Edges are treated as bidirectional (pedestrian pathways can be
            # walked in either direction).
            graph.adjacency.setdefault(e.from_node, []).append((e.to_node, e))
            graph.adjacency.setdefault(e.to_node, []).append((e.from_node, e))
        return graph

    def nearest_node(self, latitude: float, longitude: float) -> str:
        """Find the graph node closest to an arbitrary lat/lon (e.g. the
        user's current GPS position, which won't sit exactly on a node)."""
        if not self.nodes:
            raise RouteNotFoundError("Routing graph is empty.")
        best_id, best_dist = None, float("inf")
        for node in self.nodes.values():
            d = haversine_distance_meters(
                latitude, longitude, node.latitude, node.longitude
            )
            if d < best_dist:
                best_id, best_dist = node.id, d
        return best_id


class RoutingService:
    def __init__(self, graph: RoutingGraph, walking_speed_mps: float = 1.3):
        self.graph = graph
        self.walking_speed_mps = walking_speed_mps

    def find_route(
        self,
        from_latitude: float,
        from_longitude: float,
        to_node_id: str,
        accessible_only: bool = False,
    ) -> RouteResult:
        start_id = self.graph.nearest_node(from_latitude, from_longitude)
        if to_node_id not in self.graph.nodes:
            raise RouteNotFoundError(f"Destination node '{to_node_id}' not found.")

        node_ids = self._a_star(start_id, to_node_id, accessible_only)
        total_distance = self._path_distance(node_ids)
        return RouteResult(
            node_ids=node_ids,
            total_distance_meters=round(total_distance, 1),
            estimated_walking_time_seconds=round(
                estimate_walking_time_seconds(total_distance, self.walking_speed_mps),
                1,
            ),
        )

    def _heuristic(self, node_id: str, goal_id: str) -> float:
        a = self.graph.nodes[node_id]
        b = self.graph.nodes[goal_id]
        return haversine_distance_meters(a.latitude, a.longitude, b.latitude, b.longitude)

    def _a_star(
        self, start_id: str, goal_id: str, accessible_only: bool
    ) -> list[str]:
        open_heap: list[tuple[float, str]] = [(0.0, start_id)]
        came_from: dict[str, str] = {}
        g_score: dict[str, float] = {start_id: 0.0}
        visited: set[str] = set()

        while open_heap:
            _, current = heapq.heappop(open_heap)
            if current == goal_id:
                return self._reconstruct_path(came_from, current)
            if current in visited:
                continue
            visited.add(current)

            for neighbor_id, edge in self.graph.adjacency.get(current, []):
                if accessible_only and not edge.accessible:
                    continue
                tentative_g = g_score[current] + edge.distance
                if tentative_g < g_score.get(neighbor_id, float("inf")):
                    came_from[neighbor_id] = current
                    g_score[neighbor_id] = tentative_g
                    f_score = tentative_g + self._heuristic(neighbor_id, goal_id)
                    heapq.heappush(open_heap, (f_score, neighbor_id))

        raise RouteNotFoundError(
            f"No route found between '{start_id}' and '{goal_id}'"
            + (" using only accessible paths." if accessible_only else ".")
        )

    @staticmethod
    def _reconstruct_path(came_from: dict[str, str], current: str) -> list[str]:
        path = [current]
        while current in came_from:
            current = came_from[current]
            path.append(current)
        path.reverse()
        return path

    def _path_distance(self, node_ids: list[str]) -> float:
        total = 0.0
        for a_id, b_id in zip(node_ids, node_ids[1:]):
            for neighbor_id, edge in self.graph.adjacency[a_id]:
                if neighbor_id == b_id:
                    total += edge.distance
                    break
        return total
